scripts/run_portserver_test.py

```diff
@@ -0,0 +1,454 @@
+        # Their start time gets updated later on with their corresponding
+        # process' start time, data regarding which is present in
+        # '/proc/<pid>/stat' file.
+        # To test all possible branches, here we are manually changing
+        # the port pool start time to 1 rather than 0.
+        port = port_pool._port_queue.pop() # pylint: disable=protected-access
+        port.start_time = 1
+        port_pool._port_queue.append(port) # pylint: disable=protected-access
+        self.assertEqual(port_pool.num_ports(), 2)
+        with swap_get_process_start_time, swap_is_port_free:
+            returned_port = port_pool.get_port_for_process(12345)
+
+        self.assertEqual(returned_port, port1)
+
+    def test_get_port_for_process_handles_no_free_port(self) -> None:
+        port = 8181
+        swap_get_process_start_time = self.swap(
+            run_portserver, 'get_process_start_time', lambda _: 0)
+        swap_is_port_free = self.swap(
+            run_portserver, 'is_port_free', lambda _: False)
+
+        port_pool = run_portserver.PortPool() # type: ignore[no-untyped-call]
+        port_pool.add_port_to_free_pool(port)
+        self.assertEqual(port_pool.num_ports(), 1)
+        with swap_get_process_start_time, swap_is_port_free, self.swap_log:
+            returned_port = port_pool.get_port_for_process(12345)
+
+        self.assertEqual(returned_port, 0)
+        self.assertIn('All ports in use.', self.terminal_logs)
+
+    def test_port_server_request_handler_handles_invalid_request(self) -> None:
+        request_handler = run_portserver.PortServerRequestHandler((8181,)) # type: ignore[no-untyped-call]
+        request_handler.handle_port_request('abcd')
+        with self.swap_log:
+            request_handler.dump_stats()
+
+        self.assertIn('client-request-errors 1', self.terminal_logs)
+
+    def test_port_server_request_handler_handles_denied_allocations(
+            self) -> None:
+        request_handler = run_portserver.PortServerRequestHandler((8181,)) # type: ignore[no-untyped-call]
+        request_handler.handle_port_request(0)
+        with self.swap_log:
+            request_handler.dump_stats()
+
+        self.assertIn('denied-allocations 1', self.terminal_logs)
+
+    def test_port_server_request_handler_handles_no_free_ports(self) -> None:
+        request_handler = run_portserver.PortServerRequestHandler((8181,)) # type: ignore[no-untyped-call]
+        swap_get_port = self.swap(
+            run_portserver.PortPool, 'get_port_for_process',
+            lambda *unused_args: 0)
+        swap_should_allocate_port = self.swap(
+            run_portserver, 'should_allocate_port', lambda _: True)
+        with self.swap_log, swap_get_port, swap_should_allocate_port:
+            request_handler.handle_port_request(1010)
+            request_handler.dump_stats()
+
+        self.assertIn('denied-allocations 1', self.terminal_logs)
```

Comment:
Can you also check that the return value (which gets sent to the client) from `handle_port_request` is correct? Please also check this (that the return value for the client is correct) elsewhere

Reply:
Done